apps/design/design.md

```diff
@@ -0,0 +1,159 @@
+
+## Internals
+
+In this section we will be talking about internal/.
+
+### JSON Handling
+
+JSON must be handled specially in our app. The basics are, if we receive fields that our
+structs do not contain, we cannot drop them.  We must send them back to the service.
+
+To handle that, we use our own custom json package that handles this.
+
+See the design at: [Design](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/internal/json/design.md)
+
+### Backend communication
+
+Communication to the backends is done via the requests/ package. requests.Token is the client
+for all communication.
+
+requests.Token communicates via REST calls that are encapsulated in the ops/ client.
+
+## Adding A Feature
+
+This is the general way to add a new feature to MSAL:
+
+- Add the REST calls to ops.REST
+- Add the higher level manipulations to requests.Token
+- Add your logic to the app/\<client\> and access the services via your requests.Token
+
+## Notable Differences To Other Clients
+
+### Confidential applications have a user ID
+
+The MSAL caching design is rather simple. These design decisions and the fact that multiple applications in different languages can share a cache mean it cannot be easily changed.
+
+The entire cache contents of a confidential.Client is read and written on 
+almost any action to and from an external cache. 
+
+It is not clear to a user that a confidential client should be per user to prevent scaling
+problems. 
+
+We cannot change the MSAL cache design at this time, therefore it should be clear that
+confidential.Client should be done per user. This must go beyond a simple doc entry
+that can be ignored. Its great to say: "we told you in the doc", but that is AFTER a support call.
+
+To make it clear, we require the user to pass a userID field. This forces the user to read what
+a userID is and recognize that it is per user.
+
+A future design might include a parent client to Confidential that keeps track of the "users".
+This way we manage the lifecycle of the clients. The current design forces the user to deal with
+this, probably via a map. We could design an implementation with features from an LRU (to avoid
+memory growth), deletion schemantics, etc.... that is more efficient than what a normal user is
+going to implement (and eliminate each user from having to do this).
+
+or.....
+
+We could create a new caching mechanism for Confidential that doesn't have this problem.  Treat
+it differently than Public. This would be a larger migration path across all SDKs, but it is
+probably the right long term move.
+
+### Use of x509.Certificate and CertFromPEM() function
+
+The original version of this package used an thumbprint and a private key to do authorizations
+based on a certificate. But there wasn't a real way to get a thumbprint.
+
+A thumbprint is defined in the Oauth spec, which we had to track down. It is an SHA-1 hash
```

Comment:
>It is an SHA-1 hash [](start = 71, length = 19)

I believe other hashing is available in the updated spec, but we have to support sha-1 for backwards compatibility.

Reply:
We use sha-1 and this is an internal detail only.